### PR TITLE
Fix 38384 selectAll Command for Webviews (previewHTML)

### PR DESCRIPTION
**Pull request by sagea**
This is a possible solution to [Fix-38384](https://github.com/Microsoft/vscode/issues/38384)

It looks like the vscode behavior to block the default select all (cmd+a) behavior is still persisting for webviews. This can be tested by adding a keydown event listener to the webview contentWindow in the webview-pre.ts file (see screenshot below) and previewing a readme.md 
screen shot 2017-11-14 at 23 25 22

There are two possible solutions that I believe can fix this problem...
1. Don't block the default behavior when using a html preview webview.
or 
2. Modify the selectAll command to trigger an electron [<webview>.selectAll()](https://electron.atom.io/docs/api/webview-tag/#webviewselectall) method.

I chose the latter solution because of simplicity, but I feel like the proper solution would be to allow default behavior when using a webview.

Let me know what you think.

### CODE REVIEW

File: src/vs/editor/browser/controller/coreCommands.ts [hunk 1]
```diff
@@ -1698,11 +1692,19 @@ class EditorOrNativeTextInputCommand extends Command {
 			return;
 		}
 
+		let editorService: any = accessor.get(IEditorService);
+		let activeEditor = editorService.getActiveEditor && editorService.getActiveEditor();
+
+		if (activeEditor.isWebviewEditor && activeEditor.webview) {
+			activeEditor.webview.selectAll();
+			return;
+		}
```
Comment (mjbvz) [inline src/vs/editor/browser/controller/coreCommands.ts]: Looks like this was just for testing? I don't think `EditorOrNativeTextInputCommand` should know anything about `selectAll`

Comment (provAlexSage) [reply src/vs/editor/browser/controller/coreCommands.ts]: Yes, this was unintentional. I also found a better solution in general so I will be pushing that up when I iron out the kinks.

**Comment by provAlexSage** [comment]
@mjbvz I pushed up changes over the weekend. Let me know what you think

File: src/vs/workbench/parts/html/browser/webviewAction.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	constructor(
+		id: string,
+		label: string,
+		@IWorkbenchEditorService private editorService: IWorkbenchEditorService
+	) {
+		super(id, label);
+	}
+
+	public run(): TPromise<boolean> {
+		const activeEditor = <any>this.editorService.getActiveEditor();
```
Comment (mjbvz) [inline src/vs/workbench/parts/html/browser/webviewAction.ts]: Add check for undefined

**Review by mjbvz — APPROVED** [review]
Looks go to me overall. Adding a few other people to take a look

File: src/vs/workbench/parts/html/browser/webviewAction.ts (hunk 2 above)
Comment (sagea) [reply src/vs/workbench/parts/html/browser/webviewAction.ts]: done

File: src/vs/workbench/parts/html/browser/webviewAction.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	constructor(
+		id: string,
+		label: string,
+		@IWorkbenchEditorService private editorService: IWorkbenchEditorService
+	) {
+		super(id, label);
+	}
+
+	public run(): TPromise<boolean> {
+		const activeEditor : any = this.editorService && this.editorService.getActiveEditor();
```
Comment (bpasero) [inline src/vs/workbench/parts/html/browser/webviewAction.ts]: Do not use `any` here but the actual type (`HtmlPreviewPart`).

**Review by bpasero — CHANGES_REQUESTED** [review]
Looks ok to me however I notice a bug: 
* open a markdown preview
* press Cmd+A
* you are not getting a nice text selection but the entire markdown view gets selected

I guess there is something missing here when a preview opens to give the markdown editor focus.

File: src/vs/workbench/parts/html/browser/webview.contribution.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+actionRegistry.registerWorkbenchAction(
+	new SyncActionDescriptor(
+		WebviewSelectAllAction,
+		WebviewSelectAllAction.ID,
+		WebviewSelectAllAction.LABEL,
+		{ primary: KeyMod.CtrlCmd | KeyCode.KEY_A },
+		KEYBINDING_CONTEXT_WEBVIEWEDITOR_FOCUS
+	),
+	'Webview: Select All'
+);
```
Comment (jrieken) [inline src/vs/workbench/parts/html/browser/webview.contribution.ts]: 👎 don't be a workbench action, just be a command, be in html.contribution.ts like the others

**Review by jrieken — CHANGES_REQUESTED** [review]

**Comment by jrieken** [comment]
Apart from code review I wonder if this is desirable for all users of the webview? Often complex layouts are created with it and in those cases you don't want to be able to get everything selected. I wonder if this needs to be enabled by the extension/user of the preview-command?

**Comment by provAlexSage** [comment]
@jrieken I am currently working on a plugin that has a complex webview with inputs. I discovered this issue because I couldn't use select all didn't work with the textarea. I tested my plugin against this fix and it works as intended.

File: src/vs/workbench/parts/html/browser/webview.contribution.ts (hunk 4 above)
Comment (provAlexSage) [reply src/vs/workbench/parts/html/browser/webview.contribution.ts]: I will make this change and push up later tonight

**Comment by jrieken** [comment]
> I tested my plugin against this fix and it works as intended.

The question is what happens to all other extensions

**Comment by provAlexSage** [comment]
@jrieken The point of this change is to enable/simulate native behavior that the "editor.action.selectAll" command is breaking (IE: it prevents the ctrl+a event from firing in the webview). 

I feel like the intended behavior should be to allow for native functionality and I feel like we should give extensions control over this, event at the risk of small visual bugs for extensions. 

That being said, I don't think this solution would give extensions control over the select all functionality since the "ctrl+a" event is still blocked (I believe.) This can always be solved by adding the "[KEYBINDING_CONTEXT_WEBVIEWEDITOR_NOT_FOCUSED](https://github.com/Microsoft/vscode/blob/0a771c44a54579ba92a334950e387491278e52eb/src/vs/workbench/parts/html/browser/webviewEditor.ts#L28)" precondition to the [selectAll core command here](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/browser/controller/coreCommands.ts#L1725)

**Comment by sagea** [comment]
I tested the KEYBINDING_CONTEXT_WEBVIEWEDITOR_NOT_FOCUSED piece and it didn't seem to work. The ctrl+a event was being fired but the native select all functionality didn't work.

**Comment by mjbvz** [comment]
@sagea Please let us know if you are able to fix the remaining review comments by the start of next week. We will be locking down the master branch to prep for the VS Code 1.19 release then and would need to merge this PR in before that. Otherwise, this will have to wait for VS Code 1.20

**Comment by sagea** [comment]
@mjbvz I am busy this week so it will probably need to be delayed till 1.20

**Comment by mjbvz** [comment]
Hey @sagea, are you still interested in getting this PR in?

**Comment by mjbvz** [comment]
Closing this PR as inactive.

@sagea Please resubmit it if you have a chance. Definitely a feature that I'd like to see included

**Pull request closed without merging**